packages/monaco/src/browser/monaco-workspace.ts

```diff
@@ -175,14 +182,34 @@ export class MonacoWorkspace implements lang.Workspace {
     @inject(ProblemManager)
     protected readonly problems: ProblemManager;
 
+    protected _workspaceFolders: WorkspaceFolder[];
+    get workspaceFolders(): WorkspaceFolder[] {
+        return this._workspaceFolders;
+    }
+
     @postConstruct()
     protected init(): void {
         this.workspaceService.roots.then(roots => {
-            const rootStat = roots[0];
-            if (rootStat) {
-                this._rootUri = rootStat.uri;
+            this._workspaceFolders = roots.map((dir, index) => this.toWorkspaceFolder(dir.uri, index));
+            if (this._workspaceFolders.length !== 0) {
+                this.resolveReady();
+            }
+        });
+
```

Comment:
I think the control flow should be:
- (1) listen for `ready`
- (2) compute _workspaceFolders and fire first event (should be synchronous)
- (3) start listening for `onWorkspaceChanged` event, perform (2) for each event